lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```diff
@@ -0,0 +1,172 @@
+  private final long timeoutPeriod;
+  private final TimeUnit timeoutUnits;
+
+  /**
+   * Initialize a multiple asynchronous to synchronous object with a specified timeout value.
+   * @param timeoutPeriod The length of time in units given by the the timeoutUnits
+   *                      parameter before the condition automatically times out.
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the {@code release()} method.
+   * @param asyncProcessor A {@code FutureTask} object which returns {@code TAsync} that initiates the asynchronous
+   *                       processing associated with the call. This will occur inside the condition lock
+   *                       to prevent the processing from generating the signal before the calling thread blocks.
+   *                       Error conditions should be handled by throwing an exception which the caller
+   *                       will catch.
+   * @param <TAsync> The return type of the {@code asyncProcessor};
+   * @return A boolean value that indicates whether or not a timeout or error occurred.
+   * @throws InterruptedException The thread was interrupted while waiting on a condition.
+   * @throws InvalidIdentifierException The identifier parameter is invalid.
+   * @throws Exception The future task object referenced by the {@code asyncProcessor}
+   *                   parameter threw an exception.
+   */
+  public <TAsync> boolean block(final long identifier, final FutureTask<TAsync> asyncProcessor) throws Exception {
+    final boolean timeoutOccurred;
+    final ComplexCondition call = allocate();
+    try {
+      call.lock();
+      // Add the call identifier to the sleeper map so release() can identify this instantiation.
+      addSleeper(identifier, call);
+      // Invoke the caller's asynchronous processing while holding the lock
+      // so a wakeup cannot occur before the caller sleeps.
+      asyncProcessor.run();
+      // Put the caller to sleep until the ack comes back. Note: we atomically
+      // give up the look as the caller sleeps and atomically reacquire the
+      // the lock as we wake up.
+      LOG.log(Level.FINER, "Putting caller to sleep on identifier [{0}]", identifier);
+      timeoutOccurred = call.await();
+      if (timeoutOccurred) {
+        LOG.log(Level.SEVERE, "Call timed out on identifier [{0}]", identifier);
+      }
+    } finally {
+      // Whether or not the call completed successfully, always remove
+      // the call from the sleeper map, release the lock and cleanup.
+      removeSleeper(identifier);
```

Comment:
Theoretically, `removeSleeper()` can throw and the `call` condition will never get unlocked.

Reply:
There is now a second try in the finally block so that the lock will be unlocked in that case.